common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java

```diff
@@ -152,10 +178,18 @@ public long perfLogEnd(String callerName, String method) {
    * @return long duration  the difference between now and startTime, or -1 if startTime is null
    */
   public long perfLogEnd(String callerName, String method, String additionalInfo) {
-    long startTime = startTimes.getOrDefault(method, -1L);
+    String key = method;
+    if (additionalInfo != null) {
+      key = key + " - " + additionalInfo;
+    }
+    long startTime = startTimes.getOrDefault(key, -1L);
     long endTime = System.currentTimeMillis();
-    long duration = startTime < 0 ? -1 : endTime - startTime;
-    endTimes.put(method, Long.valueOf(endTime));
+    long duration = startTime < 0 ? -1 : (endTime - startTime + durations.getOrDefault(key, 0L));
+    endTimes.put(key, Long.valueOf(endTime));
+    durations.put(key, duration);
+    synchronized(querySteps) {
+      querySteps.add(key);
```

Comment:
The reason for having querySteps is to order the log entries in the same order of execution, so that it is easier to understand the flow and to debug.